tree_unique_args/src/ivt.egg

```diff
@@ -0,0 +1,52 @@
+(ruleset ivt)
+
+; IVT inverts ifs and loops when their predicates match, e.g.
+;
+;        do {
+;           if (x) Y else Z
+;        } while(x);
+; 
+;        =>
+; 
+;        if (x) {
+;            do {
+;                Y
+;            } while (x);
+;        } else {
+;            Z
+;        }
+;
+; The rules here assume that the entire loop body _is_ an if. This of course
+; will not match most loops directly: we rely on optimizations that nest
+; computations inside of an if for this rule to fire.
+
+; TODO, wait for #265 to ship
+; hmmm...
+
+;                     switch  inputs output id
+(function LiftSwitch (Expr    Expr   IdSort)     Expr)
+(rewrite (LiftSwitch inner subst outer-id) 
+         (DeepCopyExpr (SubstExpr inner subst) outer-id) :ruleset always-run)
+
+; Create demand
+(rule ((= loop (Loop id in out))
+       (ExprHasRefId loop outer-id)
+       (ExprIsValid loop)
+       (= out (All ord (Cons pred (Cons switch (Nil)))))
+       (= switch (Switch pred branches))
+       (ExprIsPure pred))
+      ((LiftSwitch switch in outer-id)) :ruleset ivt)
```

Comment:
Why do you need to separate lifting the switch from applying the following rule? Could you not lift the switch within the actions for the next rule?

Reply:
The issue here is that I only want to do the `(DeepCopy (Subst ...))` business if I have a potential match for IVT. But once I _do_ have such a match, I have to wait for DeepCopy and Subst to run so I can extract out the `thn` and `els` of the new switch.

The pattern here is to essentially run the LHS for IVT twice, once to kick off the process of lifting the switch, and a second time to create the entire new expression and union it with the loop.